File: dlitscustom/override/sales_order_commission.py
# Copyright (c) 2024, DLITS and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import flt

def validate_sales_order_commission(doc, method):
    """Validate and calculate commission for Sales Order"""
    if doc.dlits_sales_partner:
        calculate_commission(doc)

def calculate_commission(doc):
    """Calculate commission amount based on sales partner configuration"""
    if not doc.dlits_sales_partner:
        return
    
    try:
        # Get sales partner details
        sales_partner = frappe.get_doc("DLITS Sales Partner", doc.dlits_sales_partner)
        
        # Check if partner is active
        if sales_partner.status != "Active":
            frappe.throw(f"Sales Partner {sales_partner.partner_name} is not active")
        
        # Set commission type and rate from sales partner
        doc.dlits_commission_type = sales_partner.commission_type
        doc.dlits_commission_rate = sales_partner.commission_rate
        
        # Calculate commission amount
        base_amount = flt(doc.grand_total)
        
        if sales_partner.commission_type == "Percentage":
            commission_amount = base_amount * flt(sales_partner.commission_rate) / 100
        else:  # Fixed Amount
            commission_amount = flt(sales_partner.commission_rate)
        
        # Apply minimum and maximum limits
        if sales_partner.minimum_commission and commission_amount < flt(sales_partner.minimum_commission):
            commission_amount = flt(sales_partner.minimum_commission)
        
        if sales_partner.maximum_commission and commission_amount > flt(sales_partner.maximum_commission):
            commission_amount = flt(sales_partner.maximum_commission)
        
        # Set commission amount
        doc.dlits_commission_amount = commission_amount
        
        # Log commission calculation
        frappe.logger().info(f"Commission calculated for Sales Order {doc.name}: {commission_amount} for partner {sales_partner.partner_name}")
        
    except Exception as e:
        frappe.log_error(f"Error calculating commission for Sales Order {doc.name}: {str(e)}")
        frappe.throw(f"Error calculating commission: {str(e)}")

@frappe.whitelist()
def get_sales_partner_details(sales_partner):
    """Get sales partner details for commission calculation"""
    if not sales_partner:
        return {}
    
    try:
        partner = frappe.get_doc("DLITS Sales Partner", sales_partner)
        return {
            "commission_type": partner.commission_type,
            "commission_rate": partner.commission_rate,
            "minimum_commission": partner.minimum_commission,
            "maximum_commission": partner.maximum_commission,
            "status": partner.status
        }
    except Exception as e:
        frappe.log_error(f"Error getting sales partner details: {str(e)}")
        return {}

@frappe.whitelist()
def calculate_commission_amount(sales_partner, grand_total):
    """Calculate commission amount for given parameters"""
    if not sales_partner or not grand_total:
        return 0
    
    try:
        partner = frappe.get_doc("DLITS Sales Partner", sales_partner)
        base_amount = flt(grand_total)
        
        if partner.commission_type == "Percentage":
            commission = base_amount * flt(partner.commission_rate) / 100
        else:  # Fixed Amount
            commission = flt(partner.commission_rate)
        
        # Apply limits
        if partner.minimum_commission and commission < flt(partner.minimum_commission):
            commission = flt(partner.minimum_commission)
        
        if partner.maximum_commission and commission > flt(partner.maximum_commission):
            commission = flt(partner.maximum_commission)
        
        return commission
        
    except Exception as e:
        frappe.log_error(f"Error calculating commission amount: {str(e)}")
        return 0

File: dlitscustom/override/sales_order_commission.js
// Copyright (c) 2024, DLITS and contributors
// For license information, please see license.txt

// Sales Order Commission JavaScript

frappe.ui.form.on('Sales Order', {
    refresh: function(frm) {
        // Set field properties
        set_commission_field_properties(frm);
        
        // Update commission display
        update_commission_display(frm);
    },
    
    dlits_sales_partner: function(frm) {
        // When sales partner is selected, get commission details and calculate
        if (frm.doc.dlits_sales_partner) {
            get_sales_partner_commission_details(frm);
        } else {
            clear_commission_fields(frm);
        }
    },
    
    grand_total: function(frm) {
        // Recalculate commission when grand total changes
        if (frm.doc.dlits_sales_partner && frm.doc.grand_total) {
            calculate_commission_amount(frm);
        }
    }
});

function set_commission_field_properties(frm) {
    // Set commission fields as read-only
    frm.set_df_property('dlits_commission_type', 'read_only', 1);
    frm.set_df_property('dlits_commission_rate', 'read_only', 1);
    frm.set_df_property('dlits_commission_amount', 'read_only', 1);
}

function get_sales_partner_commission_details(frm) {
    if (!frm.doc.dlits_sales_partner) return;
    
    frappe.call({
        method: 'dlitscustom.override.sales_order_commission.get_sales_partner_details',
        args: {
            sales_partner: frm.doc.dlits_sales_partner
        },
        callback: function(r) {
            if (r.message) {
                let details = r.message;
                
                // Check if partner is active
                if (details.status !== 'Active') {
                    frappe.msgprint(__('Selected sales partner is not active'));
                    frm.set_value('dlits_sales_partner', '');
                    return;
                }
                
                // Set commission details
                frm.set_value('dlits_commission_type', details.commission_type);
                frm.set_value('dlits_commission_rate', details.commission_rate);
                
                // Calculate commission amount
                calculate_commission_amount(frm);
            }
        }
    });
}

function calculate_commission_amount(frm) {
    if (!frm.doc.dlits_sales_partner || !frm.doc.grand_total) return;
    
    frappe.call({
        method: 'dlitscustom.override.sales_order_commission.calculate_commission_amount',
        args: {
            sales_partner: frm.doc.dlits_sales_partner,
            grand_total: frm.doc.grand_total
        },
        callback: function(r) {
            if (r.message) {
                frm.set_value('dlits_commission_amount', r.message);
            }
        }
    });
}

function clear_commission_fields(frm) {
    frm.set_value('dlits_commission_type', '');
    frm.set_value('dlits_commission_rate', 0);
    frm.set_value('dlits_commission_amount', 0);
}

function update_commission_display(frm) {
    // Add visual indicators for commission
    if (frm.doc.dlits_commission_amount > 0) {
        frm.dashboard.add_indicator(__('Commission Amount: {0}', [format_currency(frm.doc.dlits_commission_amount)]), 'blue');
    }
    
    if (frm.doc.dlits_sales_partner) {
        frm.dashboard.add_indicator(__('Sales Partner: {0}', [frm.doc.dlits_sales_partner]), 'green');
    }
}

// Utility function to format currency
function format_currency(amount) {
    return frappe.format(amount, {fieldtype: 'Currency'});
}

// Auto-calculate commission on form load if sales partner is set
frappe.ui.form.on('Sales Order', {
    onload: function(frm) {
        if (frm.doc.dlits_sales_partner && !frm.doc.dlits_commission_amount) {
            get_sales_partner_commission_details(frm);
        }
    }
});